Fix stage out deleting unwanted files

amaltaro opened the pull request:
Attempt to fix (or mitigate) #6273 

I could not find any flaw in the code, then as discussed with Seangchan, I'm removing as much as possible file deletion commands. In summary:
- files cleanup (when failures happen) will be done mainly at top level (cleanSuccessfulStageOuts call)
- since other stage out implementation don't care for file size and checksum, I also removed it from LCGImp such that we rely solely on the exit status of lcg-cp command.
- removed deferDirectoryCreation since it was always returning False
- trying to place all wmagent exit codes in a single place

I discussed this with @hufnagel today and he does not like the removal of additional checks. However there is not much we can do here since logs are not available after the workflow is archived. Moreover, I don't expect lcg-cp returning exit status 0 when transfer faces problems (!= file size or checksum).

Pylint cleanup (and my ALAN logs) were left for a bit later. For now only sharing this code for comments/review.

amaltaro committed:
Fix stage out deleting unwanted files

fix bug

cleaned up my logs; removed checksum check

added file size and checksum back

fixed fallback stuff

raise invalid path

fix exception raised

hufnagel commented:
Ok, I went through the original discussion of the problem and looked at your changes. I don't think this has anything to do with the WMAgent code, otherwise we would have seen this with other sites. We use immediate removal of files after stageout failures for CERN for instance as xrdcp absolutely needs it. Tier0 never has seen a problem with a file being staged out and then deleted and we create and stageout an absolute ENORMOUS number of files.

I also have nothing against removing some protections if we can show that they indeed are the problem here. Making changes without understanding what is going on most likely means we don't fix the problem and remove protections that were there for a reason.

Thing is, as lcg-cp happily overwrites existing files, removing the removeCommand calls on size and checksum comparison failures makes sense, PROVIDED the last stageout retry that fails cleans up after itself somehow. If it doesn't we leave dark data at the site.

Also, why would you remove the size and checksum comparison ? We don't need to delete on them, but we should fail on them. There are reasons we in general don't trust exit codes. Also has nothing to do with this issue really.

hufnagel commented:
I have one question, is lcg-del (which seems to be the delete command used here), run synchronously or asynchronously (ie. does it delete right away or does it queue the file for deletion somewhere ) ?

If the latter it could explain the problem we are seeing...

amaltaro commented:
These commands seem to be synchronous, since we have the srm-timeout option. BTW, I'll add a 10min timeout for lcg-del commands. I couldn't find anything more direct than that.

> PROVIDED the last stageout retry that fails cleans up after itself somehow. If it doesn't we leave dark data at the site.

if the last retry fails, then a cleanup is in place (was not changed)
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WMSpec/Steps/Executors/StageOut.py#L217

> Also, why would you remove the size and checksum comparison ? We don't need to delete on them, but we should fail on them. There are reasons we in general don't trust exit codes. Also has nothing to do with this issue really.

I could add them back and only exit in case of failures, no deletion. I simply removed it because others implementation don't really care.

hufnagel commented:
I think if it is working we should leave it in. Someone spend some time making it work and we should not just remove it without a good reason. Some implementation use them and in general these checks are useful.

amaltaro commented:
Ok, I'll investigate it further this week then. But this "Someone spend some time making it work and we should not just remove it without a good reason." is not really a good argument to keep code around as well :). We know there are code in WMCore that is no longer used (or that was never used), or that has not been used during the last year or so. Thus removing them seems the most reasonable thing to do IMO, otherwise it only overcomplicates the code, makes it harder to maintain and etc (v2 of stage out plugins are a good example).

hufnagel commented:
Size and checksum comparison for stageout is in general a really good idea. We had cases where the exit code of the copy program could simply not be trusted. In addition, comparing the checksum can catch errors that the client exit code cannot tell you. Corruption of data in-flight should be rare, but cannot be detected unless you compare what went in (checksum you know the file is supposed to have) to what came out (checksum of the data as written on the other end).

amaltaro commented:
Updated the code with the file size and checksum checks again, but this time not deleting anything in place in case it fails.
Also added srm-timeout to ls and del commands, such that we remove them from srm queue in case it takes too long.

Moreover, this Alarm
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WMSpec/Steps/Executors/StageOut.py#L214
can brings us problems in case it is triggered and we don't cancel the srm requests (also leaving garbage  at the site). We could keep only the srm timeouts here, but this class is used by all plugins and then probably not a good idea.

juztas commented:
@amaltaro Just quick test with 2 sites xmls, and both failed to do fallback stageout. If 'srmv2' is used as fallback, it requires to create directories before executing cp and now this code is not calling directory creation.

amaltaro commented:
Thanks for checking this out Justas. I'd probably miss the fallback test. I'll check the rest tomorrow

ericvaandering commented:
test this please

amaltaro committed:
pylint cleanup

amaltaro commented:
@ticoann do not merge please. It still has my loggings and I'd prefer discussing it in tomorrow's meeting.

ericvaandering commented:
Needs to be rebased.

amaltaro commented:
@ticoann let's hold on this one for a bit more please. Anyways it's on the agent side only.

amaltaro commented:
Closing it in favor of #6695
